tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/trex/ConstrainedTransitionExecutor.scala

```diff
@@ -0,0 +1,111 @@
+package at.forsyte.apalache.tla.bmcmt.trex
+
+import at.forsyte.apalache.tla.bmcmt.rules.aux.Oracle
+import at.forsyte.apalache.tla.lir.oper.TlaBoolOper
+import at.forsyte.apalache.tla.lir.{BoolT1, NameEx, OperEx, TlaEx, Typed}
+import at.forsyte.apalache.tla.lir.transformations.impl.IdleTracker
+import at.forsyte.apalache.tla.lir.transformations.standard.ReplaceFixed
+
+/**
+ * ConstrainedTransitionExecutor allows us to add path constraints at certain depths.
+ *
+ * @param trex transition executor to decorate
+ * @tparam ExecutorContext executor context
+ * @author Igor Konnov
+ */
+class ConstrainedTransitionExecutor[ExecutorContext](trex: TransitionExecutor[ExecutorContext])
+    extends TransitionExecutor[ExecutorContext] {
+  type PathConstraint = List[TlaEx]
+
+  // path constraints per path length
+  private var pathConstraints: Map[Int, List[PathConstraint]] = Map.empty
+
+  /**
+   * Add a path constraint that is applied, whenever a symbolic execution is extended to the length of the path constraint.
+   * This constraint is translated into a disjunction of constraints over individual states.
+   *
+   * @param constraint path constraint
+   */
+  def addPathOrConstraint(constraint: PathConstraint): Unit = {
+    val len = constraint.length
+    pathConstraints += len -> (constraint :: pathConstraints.getOrElse(constraint.length, Nil))
+
+    assertPathOrConstraint(trex.execution, constraint)
+  }
+
+  /**
+   * Push path constraints immediately in the context.
+   */
+  def assertPathConstraints(): Unit = {
+    val exec = trex.execution
+    pathConstraints.get(exec.path.length).foreach {
+      _.foreach { cons =>
+        assertPathOrConstraint(exec, cons)
+      }
+    }
+  }
+
+  override def assumeTransition(transitionNo: Int): Unit = {
+    trex.assumeTransition(transitionNo)
+    assertPathConstraints()
+  }
+
+  override def pickTransition(): Oracle = {
+    val oracle = trex.pickTransition()
+    assertPathConstraints()
+    oracle
+  }
+
+  private def assertPathOrConstraint(execution: EncodedExecution, pathConstraint: PathConstraint): Unit = {
+    if (execution.path.length >= pathConstraint.length) {
+      val constraints =
+        execution.path.tail.zip(pathConstraint.tail).map { case ((binding, _), cons) =>
```

Comment:
A comment explaining the `.tail`s would be helpful.

Reply:
Good point. I have added a comment.